### ISSUE TITLE
empty results from match_all()

### ISSUE THREAD

**Issue by RK1912** [pyQms/pyqms#23]
HI 
I am currently running a targeted metabolomics pipeline and I wanted to use pyQms for the same. I ran it on publicly available dataset to finalize the pipeline . 
1. I first converted .raw files to .mzML files using msconvert  ( windows version ). The settings I used are SRM as spectra and MSlevels 1 to 2 
2. I then made an isotopologue library (in pyqms) with default parameters for 10 metabolites.
3. I accessed the mzML files with pymzml reader ( which worked as expected and I was able to see different fields of the file ) and I applied the match_all() to match spectra in the mzML files with the pre-defined metabolites. 

However, I am getting empty results for all spectra ( for all 197 samples ).

Is there anything I am missing in my approach? 

The following is an example of the isotope library: 
`{'C(3)H(7)N(1)O(3)': {'env': {(('N', '0.000'),): {'isot': [],
    'mass': [105.0425930894,
     106.04535444237278,
     107.04859822526103,
     108.05141269057845,
     109.04908971506664],
    'abun': [95693, 3564, 605, 21, 0],
    'relabun': [1.0,
     0.037243259830917544,
     0.006325210252071238,
     0.00022270996650129225,
     7.308437265552054e-07],
    'c_peak_pos': [0, 1, None, None, None],
    'n_c_peaks': 2.0,
    1: {'mz': [106.04986955617,
      107.05263090914278,
      108.05587469203103,
      109.05868915734844,
      110.05636618183664],
     'tmzs': [{106049, 106050}, {107052, 107053}, None, None, None],
     'atmzs': {106049, 106050, 107052, 107053}}}},
  'cc': {'C': 3, 'H': 7, 'N': 1, 'O': 3}}`

The dataset I used is here : [https://www.ebi.ac.uk/metabolights/MTBLS405/metabolites](url)

Thanks !

**Comment by JB-MS** [comment]
Hi,

thanks for using pyQms! To identify the problem I need to look into the data. In the publication I can see that they used a triple quad MS to analyze the data and unfortunately they do not state anything about resolution of the data. So my first guess here is, that pyQms is run with a higher accuracy (defaults to 5 ppm) than the data was measured. pyQms is optimized for high resolution (and untargeted) MS data, so analyzing this SRM data might be challenging. However, I will have a look into it and I'll get back to you.

Cheers,
Johannes

**Comment by JB-MS** [comment]
Hi,

I have been able to get quantitative values by adjusting some params. However, you need to optimize the parameters further in order to get more and usable results.  E.g. the quality criteria for matching are considerably lowered.

`params = {
    "LOWER_MZ_LIMIT": 50,
    "MINIMUM_NUMBER_OF_MATCHED_ISOTOPOLOGUES": 1,
    "REL_MZ_RANGE": 20e-6,
    "REQUIRED_PERCENTILE_PEAK_OVERLAP": 0.25
}`

Besides that, you have to careful evaluate the results, since pyQms is not optimal for such kind of (low res and targeted) MS data.

If you still have problems, you can also post your code her and I can help you with that as well.

Cheers,
Johannes

**Comment by RK1912** [comment]
Hi Johannes , 

I tried with the params you suggested and also lowered some other criteria for matching / reporting . But I still get empty results.
Here is my code:
`pyqms_params = {"REL_MZ_RANGE": 20e-6, "MINIMUM_NUMBER_OF_MATCHED_ISOTOPOLOGUES" : 1, "LOWER_MZ_LIMIT" : 50, "REQUIRED_PERCENTILE_PEAK_OVERLAP": 0.01, "M_SCORE_THRESHOLD" : 0.1}

molecules = [i for i in final_metab_df["formula"]]
metab_isotopes = pyqms.IsotopologueLibrary(molecules, charges=[1], params = pyqms_params, verbose=False)

all_files = glob.glob("../../mzml_files/*.mzML")

all_results = []
for i in all_files:
    mzml = pymzml.run.Reader(i)
    
    r = None
    for j in mzml:
        if j.ms_level == 2:
            r = metab_isotopes.match_all(mz_i_list=j.peaks("centroided"),spec_id=j["id"],spec_rt=j.scan_time[0], results=r)
            
    all_results.append(r)     `

Please let me know where I have gone wrong. 

Thanks , 
RK

**Comment by JB-MS** [comment]
Hi,

I have tried your code and I get quantitative values. I have some more suggestions, where things can go wrong. Have you put an '+' in front of every formula (e.g. '+C2H5NO2')? Which versions of pyQms and pymzML are you using? I have used the 'Benz1' file of their dataset, can you verify, that even this file gives no results? Can you provide me with a mzML file, which you have converted?

Cheers,
Johannes

**Comment by RK1912** [comment]
Hi , 
Thanks for your quick response and for taking time to run the code. 
I have uploaded the full code and few mzML files that I have converted here: 
[https://github.com/RK1912/Metabolomics](url)

The versions of pyqms and pymzml I am using currently are 0.6.2 and 2.4.7, respectively.

Let me know if you need anything else. I will keep trying on my side also. 

Thanks!

**Comment by JB-MS** [comment]
Hi,

your code is fine, I have tested it. The files are also fine. The main issue is that I can only quantify few molecules in e.g. the Benz1.mzML file using a very low accuracy of 500 ppm ('+C(5)H(8)O(4)','+C(4)H(8)N(2)O(3)','+C(5)H(12)N(2)O(2)','+C(5)H(5)N(5)','+C(9)H(11)N(1)O(2)').

pyQms works not optimal on such low resolution data, so I am not sure where to further adjust parameters right now. Maybe you can play around with some parameters and I give it a thought as well and get back to you.

Cheers,
Johannes

**Comment by StSchulze** [comment]
Hi,

in my experience with suboptimal quality data, changing the intensity-related parameters helps quite a bit, i.e. weighing the intensity less and allowing a larger error in the intensity. This might only be relevant for low-intensity data though, and of course could even increase the problem if the issue is on the m/z accuracy side.
Anyway, the parameters for that are:
`MZ_SCORE_PERCENTILE` (increase to weigh it more towards m/z, if I remember correctly)
`REL_I_RANGE` (increase as well to allow larger intensity error, if I remember correctly)

If you know which spectra should contain which molecules, you can also plot the theoretical isotopologues to check why it's not matching (e.g. to see if the m/z or i is the issue).

Best,
Stefan

**Comment by RK1912** [comment]
Hi 
Thanks for the help. I managed to get some results for some of the metabolites in the study. 
I also have this question - What is the minimum resolution of data for which we can use pyqms for quantification ? Or some range within which pyqms works best ? 

Thanks !

**Comment by JB-MS** [comment]
Hi,

I have never tested the limits of pyQms in terms of resolution . However I have quantified in MS2 spectra wich had a resolution of 30k. In MS1 usually 70k is used. In terms of ppm accuracy we have a default of 5, but I have also used 20 ppm or more for some data to improve the matching scores. Generally, high resolution data works best. If you have further questions, feel free to ask.

Cheers,
Johannes
